### PR TITLE
ASCN/FACETS Frontend Support

### PR DESCRIPTION
**Pull request by averyniceday**
These code changes were already reviewed and approved in PRs: #2762 #3261  #3280  #3374 
Code has been rebased to latest master.

Describe changes proposed in this pull request:
Frontend support for ASCN annotated mutation data
    
    Added new columns to Mutation Table on both Results/Patient View Pages
    - Cancer Cell Fraction
    - Clonal
    - Allele Specific Copy Number
    - ASCN Method
    - Mutant Copies
    Rich hoverover tooltips written for all new columns
    
    Numerous unit tests written for visualization of table elements and
    hoverover tooltips. Also includes unit tests for table column state
    indicators (loading/pending, error encountered)
    
    Mutation table columns are excluded/hidden when no mutation is present
    in the dataset which has a (non-empty) value set for the relevant ASCN
    property. This is accomplished by computing a property for the mutation
    table react elements holding a map from ASCN property name to a boolean
    value indicating whether any mutation present has a value for the
    relevant property. This map will initially map all properies to false,
    and will be updated/recomputed when the api request for mutation data
    completes.
    
    added/used additional symbolic string constants for api calls (see new
    enums in src/shared/constants.ts)
    NOTE TO DEVELOPERS: please adopt and expand these enums in new features
    Also, ASCN specific enums added in src/shared/lib/ASCNUtils.ts
    
    Efforts to expand type specifications in functions and variables and
    protect against non-typesafe operations
    
    Adopted use of template literals
    Some dead code was purged, and unneeded arguments were removed
    
    StandaloneMutationMapper was considered as well (to update the mutation
    table display to include the new ASCN fields.) However, it was not
    updated at this time because as currently used this feature limits
    parsed fields to a limited set of mutation attributes which do not
    include any of the ascn properties. 

## Checks
- [ ] Has tests or has a separate issue that describes the types of test that should be created. If no test is included it should explicitly be mentioned in the PR why there is no test.
- [ ] The commit log is comprehensible. It follows [7 rules of great commit messages](http://chris.beams.io/posts/git-commit/). For most PRs a single commit should suffice, in some cases multiple topical commits can be useful. During review it is ok to see tiny commits (e.g. Fix reviewer comments), but right before the code gets merged to master or rc branch, any such commits should be squashed since they are useless to the other developers. Definitely avoid [merge commits, use rebase instead.](http://nathanleclaire.com/blog/2014/09/14/dont-be-scared-of-git-rebase/)
- [ ] Is this PR adding logic based on one or more **clinical** attributes? If yes, please make sure validation for this attribute is also present in the data validation / data loading layers (in backend repo) and documented in [File-Formats Clinical data section](https://github.com/cBioPortal/cbioportal/blob/master/docs/File-Formats.md#clinical-data)!

## Any screenshots or GIFs?
If this is a new visual feature please add a before/after screenshot or gif
here with e.g. [Giphy CAPTURE](https://giphy.com/apps/giphycapture) or [Peek](https://github.com/phw/peek)

## Notify reviewers
Read our [Pull request merging
policy](../CONTRIBUTING.md#pull-request-merging-policy). It can help to figure out who worked on the
file before you. Please use `git blame <filename>` to determine that
and notify them either through slack or by assigning them as a reviewer on the PR

### CODE REVIEW

File: src/shared/enums/ASCNEnums.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+export enum ASCNAttributes {
```
Comment (zhx828) [inline src/shared/enums/ASCNEnums.ts:1]: what's the consideration have this enum as a separate file? I see other enum you added in this pr are in constants.ts

Comment (averyniceday) [reply src/shared/enums/ASCNEnums.ts:1]: Most of the changes in constants.ts were made by @sheridancbio, looping him in. But from my understanding the constants.ts were things used as args for API calls. 

ASCNEnums is specifically for the allele_specific_copy_number model in the backend.

File: src/shared/constants.ts [hunk 2]
```diff
@@ -2,3 +2,57 @@ export const MOLECULAR_PROFILE_MUTATIONS_SUFFIX = '_mutations';
 export const MOLECULAR_PROFILE_UNCALLED_MUTATIONS_SUFFIX = `${MOLECULAR_PROFILE_MUTATIONS_SUFFIX}_uncalled`;
 export const MUTATION_STATUS_GERMLINE = 'Germline';
 export const PUTATIVE_DRIVER = 'Putative_Driver';
+
+export const SAMPLE_CANCER_TYPE_UNKNOWN = 'Unknown';
+
+/* cbioportal api request arguments */
+export const enum REQUEST_ARG_ENUM {
```
Comment (alisman) [inline src/shared/constants.ts:9]: these should already be protected by typing on the methods themselves.  i.e. you can't pass an incorrect value as an argument.

File: src/shared/components/mutationTable/column/expectedAltCopies/ExpectedAltCopiesColumnFormatter.tsx [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        sampleIds: string[],
+        sampleManager?: SampleManager | null
+    ) {
+        const sampleToTotalCopyNumber: { [key: string]: string } = {};
+        const sampleToExpectedAltCopies: { [key: string]: string } = {};
+
+        // only signify NA (no FACETS analysis done) if ASCN METHOD not set
+        // else indicate indeterminte
+        for (const mutation of data) {
+            sampleToTotalCopyNumber[mutation.sampleId] = hasASCNProperty(
```
Comment (alisman) [inline src/shared/components/mutationTable/column/expectedAltCopies/ExpectedAltCopiesColumnFormatter.tsx:93]: nested ternaries are almost impossible to read. we should avoid them

File: src/shared/components/mutationTable/column/clonal/ClonalElement.tsx [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+            </pattern>
+        </defs>
+    </svg>
+);
+
+export enum ClonalColor {
+    LIMEGREEN = 'limegreen',
+    DIMGREY = 'dimgrey',
+    LIGHTGREY = 'lightgrey',
+    STRIPED = 'url(#stripePattern)',
```
Comment (alisman) [inline src/shared/components/mutationTable/column/clonal/ClonalElement.tsx:33]: interesting!  curious how to this works

File: src/pages/resultsView/ResultsViewPageStore.ts [hunk 5]
```diff
@@ -347,19 +359,19 @@ export function buildDefaultOQLProfile(
...
                 default_oql_uniq['MUT'] = true;
                 default_oql_uniq['FUSION'] = true;
                 break;
-            case 'COPY_NUMBER_ALTERATION':
+            case AlterationTypeConstants.COPY_NUMBER_ALTERATION:
                 default_oql_uniq['AMP'] = true;
                 default_oql_uniq['HOMDEL'] = true;
                 break;
-            case 'MRNA_EXPRESSION':
+            case AlterationTypeConstants.MRNA_EXPRESSION:
```
Comment (alisman) [inline src/pages/resultsView/ResultsViewPageStore.ts:370]: thanks for fixing all this!

**Review by alisman — APPROVED** [review]

**Pull request merged**